tests/kernel/lifo/lifo_usage/src/main.c

```diff
@@ -0,0 +1,422 @@
+{
+	void *ret;
+
+	ret = k_lifo_get((struct k_lifo *)p1, K_FOREVER);
+
+	/* data pushed at last should be read first */
+	zassert_equal(ret, (void *)&data[1], NULL);
+
+	ret = k_lifo_get((struct k_lifo *)p1, K_FOREVER);
+
+	zassert_equal(ret, (void *)&data[0], NULL);
+
+	k_sem_give(&start_sema);
+}
+
+static void test_lifo_nowait(void)
+{
+	k_lifo_init(&lifo);
+
+	k_sem_init(&start_sema, 0, 1);
+
+	/* put some data on lifo */
+	k_lifo_put(&lifo, (void *)&data[0]);
+
+	k_tid_t tid = k_thread_create(&tdata, tstack, STACK_SIZE,
+				      thread_entry_nowait, &lifo, NULL, NULL,
+				      K_PRIO_PREEMPT(0), 0, 0);
+
+	k_lifo_put(&lifo, (void *)&data[1]);
+
+	/* Allow another thread to read lifo */
+	k_sem_take(&start_sema, K_FOREVER);
+	k_thread_abort(tid);
+}
+
+static void thread_entry_wait(void *p1, void *p2, void *p3)
+{
+	TC_PRINT("Push &data[0]\n");
+	k_lifo_put((struct k_lifo *)p1, (void *)&data[0]);
+
+	TC_PRINT("Push &data[1]\n");
+	k_lifo_put((struct k_lifo *)p1, (void *)&data[1]);
+	k_sem_give(&wait_sema);
+}
+
+static void test_lifo_wait(void)
+{
+	int *ret;
+
+	k_lifo_init(&plifo);
+	k_sem_init(&wait_sema, 0, 1);
+
+	k_tid_t tid = k_thread_create(&tdata1, tstack1, STACK_SIZE,
+				      thread_entry_wait, &plifo, NULL, NULL,
+				      K_PRIO_PREEMPT(0), 0, 0);
+
+	TC_PRINT("Start reading lifo\n");
+
+	ret = k_lifo_get(&plifo, K_FOREVER);
+
+	zassert_equal(ret, (void *)&data[0], NULL);
+
+	TC_PRINT("take semaphore to read next data\n");
+
+	k_sem_take(&wait_sema, K_FOREVER);
+
+	ret = k_lifo_get(&plifo, K_FOREVER);
+
+	zassert_equal(ret, (void *)&data[1], NULL);
+
+	k_thread_abort(tid);
+}
+
+static bool is_timeout_in_range(u32_t start_time, u32_t timeout)
+{
+	u32_t stop_time, diff;
+
+	stop_time = k_cycle_get_32();
+	diff = SYS_CLOCK_HW_CYCLES_TO_NS(stop_time -
+					start_time) / NSEC_PER_USEC;
+	diff = diff / USEC_PER_MSEC;
+	return timeout <= diff;
+}
+
+static void test_timeout_empty_lifo(void)
+{
+	void *packet;
+
+	u32_t start_time, timeout;
+
+	timeout = 10;
+
+	start_time = k_cycle_get_32();
+
+	packet = k_lifo_get(&lifo_timeout[0], timeout);
+
+	zassert_equal(packet, NULL, NULL);
+
+	zassert_true(is_timeout_in_range(start_time, timeout), NULL);
+
+	/* Test empty lifo with timeout of K_NO_WAIT */
+	packet = k_lifo_get(&lifo_timeout[0], K_NO_WAIT);
+	zassert_equal(packet, NULL, NULL);
+}
+
+static void test_timeout_non_empty_fifo(void)
+{
+	void *packet, *scratch_packet;
+
+	/* Test k_lifo_get with K_NO_WAIT */
+	scratch_packet = get_scratch_packet();
+	k_lifo_put(&lifo_timeout[0], scratch_packet);
+	packet = k_lifo_get(&lifo_timeout[0], K_NO_WAIT);
+	zassert_true(packet != NULL, NULL);
+	put_scratch_packet(scratch_packet);
+
+	 /* Test k_fifo_get with K_FOREVER */
+	scratch_packet = get_scratch_packet();
+	k_lifo_put(&lifo_timeout[0], scratch_packet);
+	packet = k_lifo_get(&lifo_timeout[0], K_FOREVER);
+	zassert_true(packet != NULL, NULL);
+
+	put_scratch_packet(scratch_packet);
```

Comment:
why is this at the end of the test case? who is reading this from the lifo?

Reply:
Removed it.